examples/marketplace-contract/src/lib.rs

```diff
@@ -0,0 +1,595 @@
+}
+/// Calls the [balanceOf](https://proposals.concordium.software/CIS/cis-2.html#balanceof) function of the CIS2 contract. 
+/// Returns error if the returned balance < input balance (balance param).
+fn ensure_balance<
+    S: HasStateApi,
+    T: IsTokenId + Clone + Copy,
+    A: IsTokenAmount + Ord + Default + Clone + Copy + ops::Sub<Output = A>,
+>(
+    host: &mut impl HasHost<State<S, T, A>, StateApiType = S>,
+    token_id: T,
+    nft_contract_address: &ContractAddress,
+    owner: AccountAddress,
+    balance: A,
+) -> ContractResult<()> {
+    let contract_balance: A = Cis2Client::get_balance(
+        host,
+        token_id,
+        nft_contract_address,
+        Address::Account(owner),
+    )
+    .map_err(MarketplaceError::Cis2ClientError)?;
+
+    let has_balance = contract_balance.cmp(&balance).is_ge();
+    ensure!(has_balance, MarketplaceError::NoBalance);
+
+    Ok(())
+}
+
+// Distributes Selling Price, Royalty & Commission amounts.
+fn distribute_amounts<
+    S: HasStateApi,
+    T: IsTokenId + Clone + Copy,
+    A: IsTokenAmount + Clone + Copy + ops::Sub<Output = A> + Default,
+>(
+    host: &mut impl HasHost<State<S, T, A>, StateApiType = S>,
+    amount: Amount,
+    token_owner: &AccountAddress,
+    token_royalty_state: &TokenRoyaltyState,
+    marketplace_owner: &AccountAddress,
+) -> Result<(), MarketplaceError> {
+    let amounts = calculate_amounts(
+        &amount,
+        &host.state().commission,
+        token_royalty_state.royalty,
+    );
+
+    host.invoke_transfer(token_owner, amounts.to_seller)
+        .map_err(|_| MarketplaceError::InvokeTransferError)?;
+
+    if amounts
+        .to_marketplace
+        .cmp(&Amount::from_micro_ccd(0))
+        .is_gt()
+    {
+        host.invoke_transfer(marketplace_owner, amounts.to_marketplace)
+            .map_err(|_| MarketplaceError::InvokeTransferError)?;
+    }
+
+    if amounts
+        .to_primary_owner
+        .cmp(&Amount::from_micro_ccd(0))
+        .is_gt()
+    {
+        host.invoke_transfer(&token_royalty_state.primary_owner, amounts.to_primary_owner)
+            .map_err(|_| MarketplaceError::InvokeTransferError)?;
+    };
+
+    Ok(())
+}
+
+/// Callculates the amounts (Commission, Royalty & Selling Price) to be distributed
+fn calculate_amounts(
+    amount: &Amount,
+    commission: &Commission,
+    royalty_percentage_basis: u16,
+) -> DistributableAmounts {
+    let commission_amount = amount
+        .mul(commission.percentage_basis.into())
+        .quotient_remainder(MAX_BASIS_POINTS.into());
+
+    let royalty_amount = amount
+        .mul(royalty_percentage_basis.into())
+        .quotient_remainder(MAX_BASIS_POINTS.into());
+
+    DistributableAmounts {
+        to_seller: amount
+            .subtract_micro_ccd(commission_amount.0.micro_ccd())
+            .subtract_micro_ccd(royalty_amount.0.micro_ccd()),
+        to_marketplace: commission_amount.0,
+        to_primary_owner: royalty_amount.0,
+    }
+}
+
+#[cfg(test)]
```

Comment:
Use `#[cfg(concordium_cfg_test)]` so that tests are runnable using `cargo concordium`.